### PR TITLE
verify added

### PR DESCRIPTION
**Pull request by tworec**
Hey,
it's a great small crate. Thanks for it.

From its name, I guess that it is meant to support `sign` op, but `verify` is complementary, so I'm proposing to add it.

### CODE REVIEW

**Comment by tomusdrw** [comment]
Nope, I don't see why it's required. You can just do:
```
let signature = Signature { v, r, s };
let pub_key = signature.recover(message)?;
assert_eq!(pub_key.address() = address); // or check the public key itself
```

Is there an advantage that I'm missing here? What's the use case that you want to tackle and why you think that current approach is not feasible?

**Comment by tworec** [comment]
Yup! I think your reasoning is ok. I just think explicite `verify` function would be handy, especially for not experienced. Otherwise, I think users of this crate would appreciate if you include example above into comments for `recover` and/or docs.

That said, feel free to close this PR.

File: Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,8 @@ rustc-hex = "2.0"
 secp256k1 = "0.12"
 serde = { version = "1.0", features = ["derive"]}
 parity-crypto = "0.3"
+lazy_static = "1.0"
```
Comment (tomusdrw) [inline Cargo.toml]: Let's just create the context on demand.

File: src/key.rs [hunk 2]
```diff
@@ -28,15 +32,20 @@ impl fmt::Debug for Signature {
 
 impl Signature {
     /// Recover the signer of the message.
-    pub fn recover(&self, message: &[u8]) -> Result<PublicKey, secp256k1::Error> {
-        use secp256k1::{RecoverableSignature, Message, RecoveryId, Secp256k1};
+    pub fn recoverable_signature(&self) -> Result<secp256k1::RecoverableSignature, secp256k1::Error> {
```
Comment (tomusdrw) [inline src/key.rs]: Let's make it private

File: src/key.rs [hunk 3]
```diff
@@ -85,6 +94,27 @@ impl PublicKey {
     pub fn address(&self) -> &[u8; 20] {
         &self.address
     }
+
+    /// Checks if given `signature` is a valid ECDSA signature for `message` using the this public key.
+    /// Returns `Ok(true)` on success.
+    pub fn verify(self, signature: &Signature, message: &[u8]) -> Result<bool, secp256k1::Error> {
```
Comment (tomusdrw) [inline src/key.rs]: `&self` would do, I don't think we need to consume it here

File: Cargo.toml [hunk 4]
```diff
@@ -14,6 +14,8 @@ rustc-hex = "2.0"
...
 
 [dev-dependencies]
 serde_json = "1.0"
+rand = "0.6"
```
Comment (tomusdrw) [inline Cargo.toml]: Let's make the tests deterministic for now  and avoid rand, we can add quick-check later.

**Review by tomusdrw — COMMENTED** [review]
Willing to accept that after review comments are addressed.

File: Cargo.toml (hunk 1 above)
Comment (tworec) [reply Cargo.toml]: lazy static is on demand, as the name suggests
https://docs.rs/lazy_static/1.3.0/lazy_static/

File: Cargo.toml (hunk 4 above)
Comment (tworec) [reply Cargo.toml]: ok

File: src/key.rs (hunk 2 above)
Comment (tworec) [reply src/key.rs]: done

File: src/key.rs (hunk 3 above)
Comment (tworec) [reply src/key.rs]: ok

File: src/ec.rs [hunk 5]
```diff
@@ -2,100 +2,134 @@ pub use self::secp256k1::*;
 
 #[cfg(feature = "secp256k1-c")]
 mod secp256k1 {
-    /// Alias type for `secp256k1::Error`
-    pub type Error = secp256k1::Error;
+    /// `secp256k1::Error`
+    pub use secp256k1::Error;
+
+    lazy_static::lazy_static! {
```
Comment (tomusdrw) [inline src/ec.rs]: Do you mind if we remove it for this PR? I'd like to see some benchmarks that justify `lazy_static` first.

**Review by tomusdrw — APPROVED** [review]
Looks good apart from minor grumble about the context

File: src/ec.rs (hunk 5 above)
Comment (tworec) [reply src/ec.rs]: no problem

**Review by tomusdrw — APPROVED** [review]
thanks!

**Comment by tworec** [comment]
can u publish it to crates.io as v0.4.0?

**Comment by tomusdrw** [comment]
Just published, thanks for a reminder.

**Pull request merged**